tests/e2e/failover_test.go

```diff
@@ -0,0 +1,170 @@
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	_ "google.golang.org/grpc/health"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+)
+
+const (
+	// in sync with how kubernetes uses etcd
+	// https://github.com/kubernetes/kubernetes/blob/release-1.28/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L59-L71
+	keepaliveTime    = 30 * time.Second
+	keepaliveTimeout = 10 * time.Second
+	dialTimeout      = 20 * time.Second
+
+	clientRuntime  = 10 * time.Second
+	requestTimeout = 100 * time.Millisecond
+
+	// common assertion
+	expectedMinTotalRequestsCount = 300
+	// happy case assertion
+	expectedMaxFailedRequestsCount = 5
+	// negative case assertion
+	expectedMinFailedRequestsCount = 90
+)
+
+func TestFailoverOnDefrag(t *testing.T) {
+	tcs := []struct {
+		name           string
+		clusterOptions []e2e.EPClusterOption
+
+		disableClientHealthCheck bool
+		assertion                func(t *testing.T, requestVolume, successfulRequestCount int)
+	}{
+		{
+			name: "defrag failover happy case",
+			clusterOptions: []e2e.EPClusterOption{
+				e2e.WithClusterSize(3),
+				e2e.WithExperimentalStopGRPCServiceOnDefrag(true),
+				e2e.WithEnvVars(map[string]string{"GOFAIL_FAILPOINTS": `defragBeforeCopy=sleep("10s")`}),
+			},
+			assertion: assertionInHappyCase,
+		},
+		{
+			name: "defrag blocks one-third of requests with stopGRPCServiceOnDefrag set to false",
+			clusterOptions: []e2e.EPClusterOption{
+				e2e.WithClusterSize(3),
+				e2e.WithExperimentalStopGRPCServiceOnDefrag(false),
+				e2e.WithEnvVars(map[string]string{"GOFAIL_FAILPOINTS": `defragBeforeCopy=sleep("10s")`}),
+			},
+			assertion: assertionInNegativeCase,
+		},
+		{
+			name: "defrag blocks one-third of requests with stopGRPCServiceOnDefrag set to true and client health check disabled",
+			clusterOptions: []e2e.EPClusterOption{
+				e2e.WithClusterSize(3),
+				e2e.WithExperimentalStopGRPCServiceOnDefrag(true),
+				e2e.WithEnvVars(map[string]string{"GOFAIL_FAILPOINTS": `defragBeforeCopy=sleep("10s")`}),
+			},
+			disableClientHealthCheck: true,
+			assertion:                assertionInNegativeCase,
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			e2e.BeforeTest(t)
+			clus, cerr := e2e.NewEtcdProcessCluster(context.TODO(), t, tc.clusterOptions...)
+			require.NoError(t, cerr)
+			t.Cleanup(func() { clus.Stop() })
+
+			endpoints := clus.EndpointsGRPC()
+
+			requestVolume, successfulRequestCount := 0, 0
+			g := new(errgroup.Group)
+			g.Go(func() (lastErr error) {
+				dialOptions := make([]grpc.DialOption, 0)
+				if !tc.disableClientHealthCheck {
+					dialOptions = []grpc.DialOption{
+						grpc.WithDisableServiceConfig(),
+						grpc.WithDefaultServiceConfig(`{"loadBalancingPolicy": "round_robin", "healthCheckConfig": {"serviceName": ""}}`),
+					}
+				}
```

Comment:
Can you make `grpcDialOptions` a field on `tc`. Having it directly on test case will make it more readable when reading the test cases vs some custom variable that needs to be looked up.

Reply:
Make sense to me. Done.